File: samlang-service/src/main/kotlin/samlang/service/SourceProcessor.kt
package samlang.service

import samlang.ast.common.ModuleReference
import samlang.ast.common.Sources
import samlang.ast.lang.Module
import samlang.ast.mir.MidIrCompilationUnit
import samlang.checker.ErrorCollector
import samlang.checker.typeCheckSources
import samlang.compiler.asm.AssemblyGenerator
import samlang.compiler.hir.compileSources
import samlang.compiler.mir.MidIrGenerator
import samlang.errors.CompileTimeError
import samlang.optimization.Optimizer
import samlang.parser.buildModuleFromText

fun checkSources(sourceHandles: List<Pair<ModuleReference, String>>): Pair<Sources<Module>, List<CompileTimeError>> {
    val errorCollector = ErrorCollector()
    val moduleMappings = mutableMapOf<ModuleReference, Module>()
    for ((moduleReference, text) in sourceHandles) {
        val (module, parseErrors) = buildModuleFromText(moduleReference = moduleReference, text = text)
        parseErrors.forEach { errorCollector.add(compileTimeError = it) }
        moduleMappings[moduleReference] = module
    }
    val (checkedSources, _) = typeCheckSources(
        sources = Sources(moduleMappings = moduleMappings),
        errorCollector = errorCollector
    )
    return checkedSources to errorCollector.collectedErrors
}

@ExperimentalStdlibApi
fun lowerToAssemblyString(
    source: Sources<Module>,
    entryModuleReference: ModuleReference,
    optimizer: Optimizer<MidIrCompilationUnit>
): String {
    val highIrSources = compileSources(sources = source)
    val unoptimizedCompilationUnit = MidIrGenerator.generate(
        sources = highIrSources,
        entryModuleReference = entryModuleReference
    )
    val optimizedCompilationUnit = optimizer.optimize(source = unoptimizedCompilationUnit)
    return AssemblyGenerator.generate(compilationUnit = optimizedCompilationUnit).toString()
}

File: typescript-rewrite/samlang-core/src/services/source-processor.ts
import type { AssemblyProgram } from '../ast/asm/asm-program';
import type ModuleReference from '../ast/common/module-reference';
import type { Sources } from '../ast/common/structs';
import type { SamlangModule } from '../ast/lang/samlang-toplevel';
import type { MidIRCompilationUnit } from '../ast/mir';
import { typeCheckSources, GlobalTypingContext } from '../checker';
import {
  compileSamlangSourcesToHighIRSources,
  compileHighIrSourcesToMidIRCompilationUnitWithSingleEntry,
  generateAssemblyInstructionsFromMidIRCompilationUnit,
} from '../compiler';
import { CompileTimeError, createGlobalErrorCollector } from '../errors';
import { parseSamlangModuleFromText } from '../parser';
import { hashMapOf } from '../util/collections';

type CheckSourcesResult = {
  readonly checkedSources: Sources<SamlangModule>;
  readonly globalTypingContext: GlobalTypingContext;
  readonly compileTimeErrors: readonly CompileTimeError[];
};

export const checkSources = (
  sourceHandles: readonly (readonly [ModuleReference, string])[]
): CheckSourcesResult => {
  const errorCollector = createGlobalErrorCollector();
  const moduleMappings = hashMapOf(
    ...sourceHandles.map(
      ([moduleReference, text]) =>
        [
          moduleReference,
          parseSamlangModuleFromText(text, errorCollector.getModuleErrorCollector(moduleReference)),
        ] as const
    )
  );
  const [checkedSources, globalTypingContext] = typeCheckSources(moduleMappings, errorCollector);
  return { checkedSources, globalTypingContext, compileTimeErrors: errorCollector.getErrors() };
};

export const lowerSourcesToAssemblyProgram = (
  sources: Sources<SamlangModule>,
  entryModuleReference: ModuleReference,
  optimizer: (compilationUnit: MidIRCompilationUnit) => MidIRCompilationUnit
): AssemblyProgram => {
  const highIrSources = compileSamlangSourcesToHighIRSources(sources);
  const unoptimizedCompilationUnit = compileHighIrSourcesToMidIRCompilationUnitWithSingleEntry(
    highIrSources,
    entryModuleReference
  );
  const optimizedCompilationUnit = optimizer(unoptimizedCompilationUnit);
  return generateAssemblyInstructionsFromMidIRCompilationUnit(optimizedCompilationUnit);
};
